app/pages/lab-pages-editor/DataManager.jsx

```diff
@@ -0,0 +1,82 @@
+/*
+Data Manager
+Responsible for fetching, updating, and saving Workflow resources from/to
+the Panoptes service.
+ */
+
+import apiClient from 'panoptes-client/lib/api-client'
+import { useEffect, useState } from 'react'
+import { WorkflowContext } from './context.jsx'
+
+export default function DataManager ({
+  // key: to ensure DataManager renders FRESH (with states reset) whenever workflowId changes, use <DataManager key={workflowId} ... />
+  children,
+  workflowId,
+}) {
+  let initialised = false
+  const [ apiData, setApiData ] = useState({
+    workflow: null,
+    status: 'ready',
+  })
+
+  // Fetch workflow when the component loads for the first time.
+  // See notes about 'key' prop, to ensure states are reset:
+  // https://react.dev/learn/you-might-not-need-an-effect#resetting-all-state-when-a-prop-changes
+  // Also see general pattern notes:
+  // https://react.dev/reference/react/useEffect#fetching-data-with-effects
+  useEffect(async function whenComponentMounts_doInitialise () {
+    console.log('+++ DataManager: fetchWorkflow', workflowId)
+    if (initialised) return 
+    if (!workflowId) return
+
+    try {
+      setApiData({
+        workflow: null,
+        status: 'fetching'
+      })
+
+      const wf = await apiClient.type('workflows').get(workflowId)
+      if (!wf) throw new Error('No workflow')
+
+      setApiData({
+        workflow: wf,
+        status: 'ready'
+      })
+      initialised = true
+
+    } catch (err) {
+      console.error('DataManager: ', err)
+      setApiData({
+        workflow: null,
+        status: 'error'
+      })
+      initialised = true  // Don't re-run the initialisation function again
+    }
+
+    return () => { initialised = false }
+  }, [ workflowId ])
+
+  /*
+  Updates the workflow with new data.
+   */
+  function update (data) {
+    console.log('+++ TODO')
+  }
+
+  const contextData = {
+    workflow: apiData.workflow,
+    update
+  }
+
+  if (!workflowId) return (<div>ERROR: no Workflow ID specified</div>)
+  // if (!workflow) return null
+
+  return (
+    <WorkflowContext.Provider
+      value={contextData}
```

Comment:
The API client is also a state store, so be aware that your context here might differ from the workflow stored in the client, once you start to edit and update values. You can debug this by logging the client to the console and inspecting its internal caches.

```js
// get workflow 1234 from Panoptes.
let workflow = await apiClient.type('workflows').get('1234')

// get the stored workflow from the client's internal caches.
workflow = await apiClient.type('workflows').get('1234')

// make some changes.
workflow.update(changeSet)

// saving fetches a fresh copy of workflow 1234 from Panoptes, via workflow.refresh(), before saving.
workflow.save()
```

https://github.com/zooniverse/panoptes-javascript-client/blob/b2f2b8aebc3ca43321619d29c5191fd8d536a4e2/lib/json-api-client/resource.js#L56-L58

Reply:
Thanks - I'm aware that the Workflow resource has its own internal data store, which is why I'm experimenting with where this data is stored and the the associated fetch & update functions. 

That empty query argument is new, though - what on earth?

Reply:
By the way, you can listen for changes to a resource too:
```js
workflow.listen('change', HANDLER)
```
https://github.com/zooniverse/panoptes-javascript-client/tree/master/lib/json-api-client#watching-a-resource-for-changes

That will fire whenever `workflow.update()` runs.